### PR TITLE
Add `Create.Elevation`, `Query.Points` methods, update `Create.View3D`

### LINKED ISSUE #1354: Add Elevation create method
#### Description:

As in the title.

### PR DESCRIPTION
**Pull request by michal-pekacki**
### Issues addressed by this PR

Closes #1354 

- added `Create.Elevation` method to create elevation views
- added `Query.Points` method to get all bounding box points
- updated `Create.View3D` method with CropBox alignment

### CODE REVIEW

File: Revit_Core_Engine/Create/View/Viewport.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Viewport.cs:47]: I would add `RecordNote` here. Also could you please test how does this change affect the behaviour of pushing the viewport via Revit adapter? Something to be added as a test script for this PR

File: Revit_Core_Engine/Create/View/Viewport.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Viewport.cs:62]: `viewportPlacementPoint` will never be null now

File: Revit_Core_Engine/Query/Points.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Points.cs:36]:
```suggestion
        [Description("Queries corner points of BoundingBoxXYZ.")]
```

File: Revit_Core_Engine/Query/Points.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Points.cs:40]: I believe you should be transforming the points by bbox.Transform if it is not `null`

**Review by pawelbaran — CHANGES_REQUESTED** [review]

File: Revit_Core_Engine/Create/View/ViewElevations.cs
Comment (pawelbaran) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:57]: Ahh ok, got it - I would then consider adding an overload to generate a marker plus single view, i.e. `ViewElevation`, to unlock the cases where the walls are not orthogonal. Or would that not make sense?

File: Revit_Core_Engine/Create/View/Viewport.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/Viewport.cs:62]: Right, I removed this check 👍

File: Revit_Core_Engine/Create/View/Viewport.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/Viewport.cs:47]: I removed this check, if `viewportPlacementPoint` is null, `Viewport` will be placed in default position `(0, 0, 0)`. I added `RecordNote`.

File: Revit_Core_Engine/Create/View/ViewElevations.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: It is not possible to create elevation view without the elevation marker. For not orthogonal walls created elevation views should be rotated but there still must be maximum 4 of them.

Comment (pawelbaran) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: Yeah, so as discussed offline, I would consider adding a method that would create a marker + single view, i.e. for a 5-walled room you would have 5 markers and 5 elevations, one per each. Not sure if that makes sense from practical point of view, so please dismiss if it does not make sense 😃

Comment (vietle-bh) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: I think it's necessary to create 1 marker per wall because many rooms have more than 4 walls.

Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: Ok, I think have a solution for this - I will rename this method to `Create.ElevationMarker` and it will always create one marker with up to 4 elevation views. For more than 4 walls or other non-rectangular rooms we should use this method multiple times - it depends on the case we have. What do you think @pawelbaran @vietle-bh?

Comment (vietle-bh) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: How about we first check if the room boundaries form a rectangle? If yes, we create 1 marker for all 4 elevations. If not, we create 1 marker per wall, each has only 1 elevation on it :bowtie:

Comment (FraserGreenroyd) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: > How about we first check if the room boundaries form a rectangle? If yes, we create 1 marker for all 4 elevations. If not, we create 1 marker per wall, each has only 1 elevation on it :bowtie:

I think that check will add unnecessary overhead particularly when calling the method multiple times.

Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewElevations.cs:56]: I think there should be no `Room` or `Space` outline input in this method.

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Elevation.cs:47]: Marek 😶‍🌫️

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Elevation.cs:46]: Wondering if it would not be worth splitting offset into 3 values: height, width and depth?

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Elevation.cs:74]:
```suggestion
        [Description("Creates elevation view referenced do elevation marker in the current Revit file.")]
```

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Elevation.cs:75]:
```suggestion
        [Input("elevationMarker", "Elevation marker to which elevation view will be referenced.")]
```

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/Elevation.cs:102]: Wouldn't it be worth for this method to be public?

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/View3D.cs:60]: Is it always needed? Or only when there is a cropbox? Will it not crash the existing code?

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/Elevation.cs:102]: Done 👍

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/View3D.cs:60]: It is always needed and cropbox has to be active to get the correct result. Without alignment 3DView looks like this:

image

File: Revit_Core_Engine/Create/View/Elevation.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/Elevation.cs:46]: I changed `offset` to work only for height and width. Depth is defined as separate input.

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check code-compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check code-compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:56]: You need to check if the line is horizontal (dot product with global Z < tolerance), if not then warning (note?) plus making the line horizontal - without that you may get wrong `angle` value

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:60]: I would check against other rotations as well - see suggestion below (but written off the top of my head, testing needed plus probably can be done smarter). Alternatively you can check based on dot product with global X and Y, the code should be simpler then?
```suggestion
            double absAngle = Math.Abs(angle);
            double refAngle = 0;
            while (refAngle - Tolerance.Angle <= absAngle)
            {
                if (Math.Abs(absAngle - refAngle) < Tolerance.Angle)
                    {
                        if (angle > 0)
                            angle = refAngle;
                        else
                            angle = -refAngle;
                        
                        break;
                    }
                
                refAngle += Math.PI;
            }
```

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:63]: Not overly relevant, but you can use an unbound line as an axis
```suggestion
            var axis = Line.CreateUnound(elevationMarkerLocation, XYZ.BasisZ);
```

File: Revit_Core_Engine/Modify/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Modify/Elevation.cs:51]: I am not a big fan of the method name here - couldn't it be `SetElevationProperties` or similar?

File: Revit_Core_Engine/Modify/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Modify/Elevation.cs:93]: redundant

File: Revit_Core_Engine/Modify/Elevation.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Modify/Elevation.cs:40]: Would be great if you could explain what does the method do - _modify_ can have a very broad meaning

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Sorry for yet another change request @michal-pekacki 🙈

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:51]: As discussed online - I will remove `ElevationMarker` output from this method.

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:56]: Done 👍

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs]: I fixed angle calculation 👍

**Review by travispotterBH — COMMENTED** [review]
In general a big fan of explicit typing (although I am guilty myself of using var at times)

File: Revit_Core_Engine/Create/View/ElevationMarker.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ElevationMarker.cs:47]: Done 👍

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/View3D.cs:295]: Done 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/View3D.cs:59]: It is always safer to use some tolerance
```C#
                if (Math.Abs(offset) > tolerance)
```

File: Revit_Core_Engine/Query/Perimeter.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Perimeter.cs:50]: Why are you changing the methodology of extracting the perimeter here? Is there a bug in the current algorithm? I am asking because we need to be very careful with this method as is it used by multiple convert methods, which means the change will require significant testing and probably agreement from S&P.

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/View3D.cs:59]: Safety first 👍

File: Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/ViewSection/ElevationByLine.cs:63]: Done 👍

File: Revit_Core_Engine/Query/Perimeter.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Query/Perimeter.cs]: I restored this method to previous version.

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/View3D.cs]: Updated 👍

**Review by pawelbaran — COMMENTED** [review]
I am happy with the code changes based on reading the code - good to merge from my perspective as soon as we get tester's approval.

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check compliance

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/View3D.cs:56]: Sorry one more comment: why did you add an `if` statement here? You are effectively culling the warning when the detail level input is undefined, which is undesirable I would say.

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/View3D.cs:46]:
```suggestion
        [Input("offset", "Offset that will be added to each side of the input BoundingBoxXYZ.")]
```

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [inline Revit_Core_Engine/Create/View/View3D.cs:75]: This will crash if there is no element under `viewTemplateId`.

```suggestion
                if ((document.GetElement(viewTemplateId) as View)?.IsTemplate != true)
```

...also not sure why did you move this part out of the try/catch clause if you are raising same warning for both cases anyways.

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Last few comments, I also noticed that you have boilerplate around setting view template (the code below being repeated a few times across different view creation method). Wouldn't it be worth centralising this under `SetViewTemplate`, same as we did with `SetViewName`?
image

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Last few comments, I also noticed that you have boilerplate around setting view template (the code below being repeated a few times across different view creation method). Wouldn't it be worth centralising this under `SetViewTemplate`, same as we did with `SetViewName`?
image

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (michal-pekacki) [reply Revit_Core_Engine/Create/View/View3D.cs:56]: Agreed `if` is not needed here 👍

File: Revit_Core_Engine/Create/View/View3D.cs
Comment (pawelbaran) [reply Revit_Core_Engine/Create/View/View3D.cs]: But isn't it worth dropping a more specific error then? Now you only avoid one `catch`, but that does not change anything else than that, right?

**Review by pawelbaran — COMMENTED** [review]
Happy with the changes now - waiting for the testers to approve the functionality + the bot checks to pass 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check versioning

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Comment by michal-pekacki** [comment]
@BHoMBot check required

**Review by pawelbaran — APPROVED** [review]
Happy to approve following the code review and testing done in collaboration with @michal-pekacki 👍

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Comment by michal-pekacki** [comment]
@BHoMBot check ready-to-merge

**Pull request closed without merging**
